snuba/replacer.py

```diff
@@ -31,8 +39,43 @@ def __init__(self, storage: WritableTableStorage, metrics: MetricsBackend) -> No
         ), f"This storage writer does not support replacements {storage.get_storage_key().value}"
         self.__replacer_processor = processor
         self.__database_name = storage.get_cluster().get_database()
-        self.__table_name = (
-            storage.get_table_writer().get_schema().get_local_table_name()
+
+    def __get_write_connections(self, replacement: Replacement) -> WriteConnections:
+        """
+        Some replacements need to be executed on each storage node of the
+        cluster instead that through a query node on distributed tables.
+        This happens when the number of shards changes and specific rows
+        resolve to a different node than they were doing before.
+
+        example: writing the tombstone for an event id. When we change the
+        shards number the tombstone may end up on a different shard than
+        the original row.
+
+        This returns connections to one replica for each shard so that,
+        in these cases, we can run the INSERT on each node.
+        """
+        query_table_name = self.__replacer_processor.get_schema().get_table_name()
+        local_table_name = self.__replacer_processor.get_schema().get_local_table_name()
+        cluster = self.__storage.get_cluster()
+
+        query_connection = cluster.get_query_connection(
+            ClickhouseClientSettings.REPLACE
+        )
+        write_every_node = replacement.write_every_node()
+        if not write_every_node or cluster.is_single_node():
+            return WriteConnections(
+                query_table_name, [(query_connection.get_host(), query_connection)]
+            )
+
+        pools = [
+            cluster.get_node_connection(ClickhouseClientSettings.REPLACE, node)
+            for node in self.__storage.get_cluster().get_local_nodes()
```

Comment:
@lynnagara should I pick the local nodes or distributed nodes here? From the Clickhouse Cluster docstring seems we should use the local nodes as they are the write nodes. 
Then I would assume we will have to properly add config for each cluster independently in our prod config. Now we have one cluster for all connections. Am I right ?

Reply:
We probably don't want to pick all the local nodes as we'd be sending the write to each replica individually. I guess my question would be who will be deciding which replica gets written to?

Reply:
Line 73 picks only the first replica of each shard. 
I am trying to find a more robust way to do load balancing, though as of this we send the queries to the first node of each shard, which is where we send normally our INSERT queries, but that would not be needed here.

Reply:
What pisses me off the most is that this could be much easier 
https://clickhouse.tech/docs/en/operations/settings/settings/#insert_shard_id
but that setting was introduced in version 21.3

Reply:
I think that's fine, though ideally we would have retry logic that would attempt to insert to another replica if the first one fails. Also I think we can avoid hard coding the replica value of `1`, how about picking `min(node.replica)` so if `1` is unavailable another one can be selected.

Reply:
This is a good point, will switch to min. Though I suspect Clickhosue would return all the configured nodes even if they are not up and running.
Will see what I can do about the retry on a different replica. That requires changes to the way we run the query in robust mode.

Reply:
Done, now we pick up to 3 replicas per shard and we try on different nodes upon failure.

Reply:
did we actually switch to min? Or is it random?

Reply:
It will address nodes in the same order as clickhouse returns them. Which is not really specified but as of now the min node will come first. 
I want to eventually test if it is worth reshuffle the list to make that random, but then each replacer will have to keep a more connections open, so it would be a tradeoff. I think we can start with this behavior.

Reply:
Actually thinking over that, I added a cache for the nodes list since `get_local_nodes` runs the query every time.
Now we refresh the list only every 10 seconds.

Reply:
Good call